src/cat-data-service/src/service/utilities/middleware/tracing_mw.rs

```diff
@@ -0,0 +1,82 @@
+use std::time::Instant;
+
+use tracing::{Instrument, Level};
+
+use poem::{
+    async_trait, web::RealIp, Endpoint, FromRequest, IntoResponse, Middleware, PathPattern,
+    Request, Response, Result,
+};
+
+/// Middleware for [`tracing`](https://crates.io/crates/tracing).
+#[derive(Default)]
+pub struct Tracing;
+
+impl<E: Endpoint> Middleware<E> for Tracing {
+    type Output = TracingEndpoint<E>;
+
+    fn transform(&self, ep: E) -> Self::Output {
+        TracingEndpoint { inner: ep }
+    }
+}
+
+/// Endpoint for `Tracing` middleware.
+pub struct TracingEndpoint<E> {
+    inner: E,
+}
+
+#[async_trait]
+impl<E: Endpoint> Endpoint for TracingEndpoint<E> {
+    type Output = Response;
+
+    async fn call(&self, req: Request) -> Result<Self::Output> {
+        let remote_addr = RealIp::from_request_without_body(&req)
+            .await
+            .ok()
+            .and_then(|real_ip| real_ip.0)
+            .map(|addr| addr.to_string())
+            .unwrap_or_else(|| req.remote_addr().to_string());
+
+        let span = tracing::span!(
+            target: module_path!(),
+            Level::INFO,
+            "request",
+            remote_addr = %remote_addr,
+            version = ?req.version(),
+            method = %req.method(),
+            uri = %req.original_uri(),
+        );
+
+        if let Some(path_pattern) = req.data::<PathPattern>() {
+            span.record("path_pattern", path_pattern.0.as_ref());
+        }
+
+        async move {
```

Comment:
What is the advantage of using tracing Span here ?

Reply:
Tracing spans accumulate data.
They also log within a spans context.  So an `info!` or `warn!` inside the span shows the data it currently knows from the span and allows you to identify why the log occurred and its context without passing the context all over the place and redundantly placing it in every log message.

Example of logging with the span enabled (these are single lines in the logs, I formatted for readability):

The error log was just `error!("liveness check - error")`, but the span lets the two logs be associated.
A more practical example, if you have a `warn!` inside an `eventdb` method that is used from multiple endpoints,  the spans will let you know precisely what endpoint caused it.

Because spans can be nested,  I have set up the logs to show the nested spans, so all data from any nested span is captured, no matter where the log occurs.

Key pieces of data here:
* `conn` field which allows all logs from a single connection to be associated with each other.
* `cleint` allows all logs caused by the same client IP (anonymized) to be associated with each other.  So we can also show larger sessions of connections and interactions, and filter out all other connections happening at the same time.

```json
{
    "timestamp": "2023-09-07T08:19:28.739489325Z",
    "level": "ERROR",
    "fields": {
        "message": "liveness check - error"
    },
    "target": "cat_data_service::service::api::health::live_get",
    "filename": "src/cat-data-service/src/service/api/health/live_get.rs",
    "line_number": 39,
    "span": {
        "client": "c79f5b9f-7009-46ba-8b84-f69011026a6a",
        "conn": "01H9QBK3B3P2SM00AX2TJS1A3E",
        "method": "GET",
        "path": "/api/health/live",
        "version": "HTTP/1.1",
        "name": "request"
    },
    "spans": [
        {
            "client": "c79f5b9f-7009-46ba-8b84-f69011026a6a",
            "conn": "01H9QBK3B3P2SM00AX2TJS1A3E",
            "method": "GET",
            "path": "/api/health/live",
            "version": "HTTP/1.1",
            "name": "request"
        }
    ],
    "threadName": "tokio-runtime-worker",
    "threadId": "ThreadId(2)"
}
{
    "timestamp": "2023-09-07T08:19:28.739574687Z",
    "level": "INFO",
    "fields": {
        "message": "close",
        "time.busy": "174µs",
        "time.idle": "5.11µs"
    },
    "target": "Endpoint",
    "filename": "src/cat-data-service/src/service/utilities/middleware/tracing_mw.rs",
    "line_number": 215,
    "span": {
        "client": "c79f5b9f-7009-46ba-8b84-f69011026a6a",
        "conn": "01H9QBK3B3P2SM00AX2TJS1A3E",
        "cpu_time_ms": 0.101,
        "duration_ms": 0.109,
        "endpoint": "/api/health/live",
        "method": "GET",
        "oid": "healthLive",
        "path": "/api/health/live",
        "status": 204,
        "version": "HTTP/1.1",
        "name": "request"
    },
    "spans": [],
    "threadName": "tokio-runtime-worker",
    "threadId": "ThreadId(2)"
}
```